broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java

```diff
@@ -0,0 +1,135 @@
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+
+class ZeebeDbPartitionTransitionStepTest {
+
+  TestPartitionTransitionContext transitionContext = new TestPartitionTransitionContext();
+
+  private final StateController stateController = mock(StateController.class);
+  private final ZeebeDb zeebeDb = mock(ZeebeDb.class);
+
+  private ZeebeDbPartitionTransitionStep step;
+
+  @BeforeEach
+  void setup() {
+
+    when(stateController.openDb()).thenReturn(zeebeDb);
+    transitionContext.setStateController(stateController);
+
+    step = new ZeebeDbPartitionTransitionStep();
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldDoNothing")
+  void shouldNotCloseZeebeDb(final Role currentRole, final Role targetRole) throws Exception {
+    // given
+    transitionTo(currentRole);
+    Mockito.clearInvocations(stateController);
+
+    // when
+    step.prepareTransition(transitionContext, 1, targetRole);
+
+    // then
+    assertThat(transitionContext.getZeebeDb()).isNotNull();
+    verify(stateController, never()).closeDb();
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldDoNothing")
+  void shouldNotReInstallZeebeDb(final Role currentRole, final Role targetRole) throws Exception {
+    // given
+    transitionTo(currentRole);
+    Mockito.clearInvocations(stateController);
+
+    // when
+    transitionTo(targetRole);
+
+    // then
+    assertThat(transitionContext.getZeebeDb()).isNotNull();
+    verify(stateController, never()).closeDb();
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldCloseZeebeDb")
+  void shouldCloseExistingZeebeDb(final Role currentRole, final Role targetRole) throws Exception {
+    // given
+    transitionTo(currentRole);
+    Mockito.clearInvocations(stateController);
+
+    // when
+    step.prepareTransition(transitionContext, 1, targetRole);
+
+    // then
+    assertThat(transitionContext.getZeebeDb()).isNull();
+    verify(stateController).closeDb();
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideTransitionsThatShouldReInstallZeebeDb")
+  void shouldInstallZeebeDb(final Role currentRole, final Role targetRole) throws Exception {
+    // given
+    if (currentRole != null) {
+      transitionTo(currentRole);
+    }
+    Mockito.clearInvocations(stateController);
+
+    // when
+    transitionTo(targetRole);
+
+    // then
+    assertThat(transitionContext.getZeebeDb()).isNotNull();
+    verify(stateController).openDb();
+  }
+
+  private static Stream<Arguments> provideTransitionsThatShouldDoNothing() {
+    return Stream.of(
+        Arguments.of(Role.CANDIDATE, Role.FOLLOWER),
+        Arguments.of(Role.FOLLOWER, Role.CANDIDATE),
+        Arguments.of(Role.CANDIDATE, Role.LEADER));
+  }
+
+  private static Stream<Arguments> provideTransitionsThatShouldReInstallZeebeDb() {
+    return Stream.of(
+        Arguments.of(null, Role.FOLLOWER),
+        Arguments.of(null, Role.LEADER),
```

Comment:
I think this should also happen on `null -> Role.CANDIDATE`

I guess this is not possible in a Raft state machine. But we could have a weird timing in which we missed the transition to follower, or an error occurred on that transition. And then we would see a direct transition from null to candidate.

Thinking more about it, this might be a broader issue. Right now the tests cover transitions as per the Raft state machine. But we could also see transitions not possible in the Raft state machine.

For example:

1. Raft transition: CANDIDATE -> LEADER / Zeebe transition from FOLLOWER to LEADER succeeds
2. Raft transition: LEADER -> FOLLOWER / Zeebe transition from LEADER to FOLLOWER has an error, therefore `setCurrentRole` is never called. In this case the Zeebe partition would still report `currentRole == LEADER` becasue it was the last successful transition
3. Raft transition FOLLOWER-> CANDIDATE / Zeebe transition from LEADER to CANDIDATE 

It's maybe a bad example, because CANDIDATE doesn't mean much for us. But what I do want to show is that more transitions are possible for the steps then what Raft state model allows.

Reply:
Errors in installation are a bit more problematic, I think. If steps are partially installed there might be some inconsistency. If LEADER -> FOLLOWER transition failed in Zeebe, I think we should never do "Zeebe transition from LEADER to CANDIDATE". If a transition fails in between, we should probably close all, may be transition to INACTIVE to ensure that we restart with a clean state. But that is a different topic. Here for the sake of completeness, I will add tests for the all possible transitions. Although I have to think what should be the semantics of those.